typing/typing_recovery.ml

```diff
@@ -0,0 +1,178 @@
+  match !ref_errors with
+  | Some (_, h) -> RawTypeHash.mem h te
+  | _ -> false
+
+let rec erroneous_expr_check e =
+  erroneous_type_check e.Typedtree.exp_type
+  || match e.Typedtree.exp_desc with
+  | Typedtree.Texp_ident (p, _, _) when Ident.name (Path.head p) = "_" -> true
+  | Typedtree.Texp_apply (e', _) -> erroneous_expr_check e'
+  | _ -> false
+
+let erroneous_type_register te =
+  let te = Types.Transient_expr.coerce te in
+  match !ref_errors with
+  | Some (_, h) -> RawTypeHash.replace h te ()
+  | None -> ()
+
+let with_saved_types ?warning_attribute ?save_part f =
+  (* Save the current global state, then reset it to empty,
+     so it can performs [f] in a clean environment. *)
+  let saved_types = Cmt_format.get_saved_types () in
+  Cmt_format.set_saved_types [];
+  try
+    (* When [f] succeed, our "snapshot" is dropped. Just [saved_types]
+       from [f] are preserved. *)
+    let result = match warning_attribute with
+      | None -> f ()
+      | Some attr -> Builtin_attributes.warning_scope attr f
+    in
+    begin
+      match save_part with
+      | None ->
+          (* Succeed but there is no saved parts, all local
+             fragment are discared.*)
+          ()
+      | Some f ->
+          (* Succeed and there is saved parts, we add the typedtree
+             fragment to the global saved types. *)
+          Cmt_format.set_saved_types (f result :: saved_types)
+    end;
+    result
+  with exn ->
+    (* We ensure that no information is lost by merging partial
+       information with information previously saved before
+       resetting the environment. *)
+    let partial_saved_types = Cmt_format.get_saved_types () in
+    Cmt_format.set_saved_types (partial_saved_types @ saved_types);
+    reraise exn
+
+module Saved_parts = struct
+  let attribute = Location.mknoloc "ocaml.saved-parts"
+
+  module H = Ephemeron.K1.Make(
+    struct
+      type t = string
+      let hash = Hashtbl.hash
+      let equal = String.equal
+    end)
+
+  let table = H.create 7
+
+  let gensym =
+    let counter = ref 0 in
+    fun () -> incr counter; !counter
+
+  let store parts =
+    let id = string_of_int (gensym ()) in
+    let key = Parsetree.Pconst_integer (id, None) in
+    H.add table id parts;
+    key
+end
+
+let flush_saved_types () =
+  match Cmt_format.get_saved_types () with
+  | [] -> []
+  | parts ->
+    Cmt_format.set_saved_types [];
+    let open Ast_helper in
+    let pconst_desc = Saved_parts.store parts in
+    let pexp = Exp.constant { pconst_desc; pconst_loc = !default_loc } in
+    let pstr = Str.eval pexp in
+    [ Attr.mk Saved_parts.attribute (Parsetree.PStr [ pstr ]) ]
+
+let incorrect_attribute =
+  Ast_helper.Attr.mk (Location.mknoloc "ocaml.incorrect") (Parsetree.PStr [])
+
+let recovery_attributes attrs =
+  let attrs' = incorrect_attribute :: flush_saved_types () in
+  match attrs with
+  | [] -> attrs'
+  | attrs -> attrs' @ attrs
+
+module Error_set = Set.Make (struct
```

Comment:
The use of a set might be correct ... but I have an uneasy feeling about it.
For one thing it seems heavy weight (do we really have that many errors that it makes sense to build a set instead of checking for membership before inserting in a list?) but also I have this vague sensation that it is just trying to sweep an fundamental issue under the rug: we're trying to hide duplicated errors, when we shouldn't have duplicated errors.

I could very well be incorrect: perhaps duplicated errors are unavoidable. But I'd like to be convinced of that fact.
For instance: I pointed out duplicated errors in my previous review, this new code hides them, but could I get an explanation of why they were present in the first place?